File: backend/src/routes/recipes.js
const { addAudit, machineById } = require("../domain");
const { parseJsonBody, requireFields, sendJson } = require("../http-utils");

async function handleRecipeRoutes(req, res, pathname, store) {
  if (req.method === "POST" && pathname === "/api/recipes") {
    const body = await parseJsonBody(req);
    requireFields(body, ["equipmentId", "name", "version", "parameters"]);
    const payload = await store.update((state) => {
      if (!machineById(state, body.equipmentId)) {
        const error = new Error("Equipment not found");
        error.statusCode = 404;
        throw error;
      }
      const id = `RCP-${String(state.recipeSeq).padStart(3, "0")}`;
      state.recipeSeq += 1;
      state.recipes.unshift({
        id,
        equipmentId: body.equipmentId,
        name: String(body.name).trim(),
        version: String(body.version).trim(),
        parameters: String(body.parameters).trim(),
        active: true
      });
      addAudit(state, `${id} Recipe 已建立`, body.actor || "System Admin");
      return { message: `${id} 已新增` };
    });
    sendJson(res, 201, payload);
    return true;
  }

  return false;
}

module.exports = {
  handleRecipeRoutes
};

File: backend/app/routes/recipes.py
from typing import Any
from urllib.parse import unquote

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse

from ..auth import require_roles
from ..http_utils import body_text, read_json_body, require_fields
from ..services import recipe_service

router = APIRouter(prefix="/api/recipes", tags=["recipes"])


@router.post("", response_model=None)
async def create_recipe(request: Request) -> JSONResponse:
    require_roles(request.state.user, "admin")
    body = await read_json_body(request)
    require_fields(body, ["equipmentId", "name", "version", "parameters"])
    result = recipe_service.create_recipe(
        request.app.state.store,
        equipment_id=str(body["equipmentId"]),
        name=body_text(body, "name"),
        version=body_text(body, "version"),
        parameters=body_text(body, "parameters"),
        actor=str(body.get("actor") or "System Admin"),
    )
    return JSONResponse(result, status_code=201)


@router.post("/{recipe_id}/deactivate")
async def deactivate_recipe(recipe_id: str, request: Request) -> dict[str, Any]:
    require_roles(request.state.user, "admin")
    body = await read_json_body(request)
    return recipe_service.deactivate_recipe(
        request.app.state.store,
        recipe_id=unquote(recipe_id),
        actor=str(body.get("actor") or "System Admin"),
    )
